### PR TITLE
Sidecar judge which component can be process by app configuration

### PR DESCRIPTION
**Pull request by theonefx**
# Description

This PR provide an optional way for daprd to judge which module should process. If app provide a module name list, daprd will only init module which in the list.
By default, daprd will process all component which can be loaded.

## Issue reference

Please reference the issue this PR will close: https://github.com/dapr/dapr/issues/2940

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Unit tests passing
* [X] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -400,7 +401,7 @@ func (a *DaprRuntime) initRuntime(opts *runtimeOpts) error {
 	a.daprHTTPAPI.SetActorRuntime(a.actor)
 	grpcAPI.SetActorRuntime(a.actor)
 
-	// TODO: Remove feature flag once feature is ratified
+	// TODO: Remove feature flag once feature is l
```
Comment (daixiang0) [inline pkg/runtime/runtime.go:404]: What is it for?

File: pkg/runtime/runtime.go [hunk 2]
```diff
@@ -387,6 +386,8 @@ func (a *DaprRuntime) initRuntime(opts *runtimeOpts) error {
 
 	a.loadAppConfiguration()
 
+	go a.processComponents()
```
Comment (daixiang0) [inline pkg/runtime/runtime.go:389]: What is it for?

Comment (theonefx) [reply pkg/runtime/runtime.go:389]: after get app configuration, then start process components. so runtime can judge which module can be process.

File: pkg/runtime/runtime.go (hunk 1 above)
Comment (theonefx) [reply pkg/runtime/runtime.go:404]: my mistake, no need change this

**Comment by yaron2** [comment]
@theonefx tests are failing.

**Comment by theonefx** [comment]
> @theonefx tests are failing.

I'll fix it right away

**Comment by theonefx** [comment]
@yaron2 @daixiang0 the e2e test has passed

**Comment by artursouza** [comment]
Please, validate this with E2E tests first. Overall, I am concerned about this feature changing the order of sidecar initialization. 

How is this different than https://docs.dapr.io/operations/components/component-scopes/ ?

**Review by artursouza — CHANGES_REQUESTED** [review]
Please, compare advantages of this vs https://docs.dapr.io/operations/components/component-scopes/ ?

**Comment by theonefx** [comment]
> Please, validate this with E2E tests first. Overall, I am concerned about this feature changing the order of sidecar initialization.
> 
> How is this different than https://docs.dapr.io/operations/components/component-scopes/ ?


There is a key difference: "Who decides which components can be loaded" :
1. Namespace-based solutions defer this decision until the deployment phase;
2. AppConfig-based solution, this is decided at the development phase;

The difference is that developers or operations maintain component data;

If AppConfig-based solution is supported, the following scenario can be implemented:
Operations can maintain a large component pool, and application developers can determine which components are loaded just by looking at component names.

Of course, AppConfig-based solution is an optional feature, and Namespace-based is valid whether or not developers configure it;

**Comment by artursouza** [comment]
> > Please, validate this with E2E tests first. Overall, I am concerned about this feature changing the order of sidecar initialization.
> > How is this different than https://docs.dapr.io/operations/components/component-scopes/ ?
> 
> There is a key difference: "Who decides which components can be loaded" :
> 
> 1. Namespace-based solutions defer this decision until the deployment phase;
> 2. AppConfig-based solution, this is decided at the development phase;
> 
> The difference is that developers or operations maintain component data;
> 
> If AppConfig-based solution is supported, the following scenario can be implemented: Operations can maintain a large component pool, and application developers can determine which components are loaded just by looking at component names.
> 
> Of course, AppConfig-based solution is an optional feature, and Namespace-based is valid whether or not developers configure it;


My main question is about the component scope section. What is the added value of this changed compared to the existing filter we have already implemented here: https://github.com/dapr/dapr/blob/0ba79334e6fe81e2a0539b70abfe2b696beef05f/pkg/runtime/runtime.go#L1688

In summary, this feature is already present in runtime. Unless I am missing the context here.

**Comment by berndverst** [comment]
Let me summarize my understanding:

The current approach lets the operator / devops person define which Dapr components exist in a certain namespace and which applications have access to the component.

But additionally an application developer may want to improve the performance of their application (e.g. limit the memory footprint of the sidecar, remove components that are slow to initialize and not needed etc) from the sidecar that connects to their application.

The request here is to **additionally instruct the sidecar only to initialize certain components, even if the operator allowed the application to have access to more components**.

I can see this being desirable in organizations where developers cannot manage / modify component definitions themselves.

**Pull request closed without merging**